plenum/test/node_catchup_with_3pc/test_pp_obsolescence_check_fail_for_delayed.py

```diff
@@ -0,0 +1,57 @@
+import pytest
+import asyncio
+
+from plenum.test import waits
+from plenum.test.delayers import ppDelay, pDelay, cDelay
+from plenum.test.helper import sdk_send_random_and_check
+from plenum.test.node_request.test_timestamp.helper import get_timestamp_suspicion_count
+from plenum.test.node_catchup.helper import ensure_all_nodes_have_same_data
+from plenum.test.stasher import delay_rules
+from plenum.test.spy_helpers import get_count
+from stp_core.loop.eventually import eventually
+
+
+# TODO better to use non hard coded value of nodes number
+# but for now it would be tricky since tconf can't depend on txnPoolNodeSet
+
+# should be big enough to pass PP during normal ordering flow
+PATCHED_ACCEPTABLE_DEVIATION_PREPREPARE_SECS = waits.expectedPrepareTime(4)
+
+
+@pytest.fixture(scope="module")
+def tconf(tconf):
+    old_value = tconf.ACCEPTABLE_DEVIATION_PREPREPARE_SECS
+    tconf.ACCEPTABLE_DEVIATION_PREPREPARE_SECS = PATCHED_ACCEPTABLE_DEVIATION_PREPREPARE_SECS
+    yield tconf
+    tconf.ACCEPTABLE_DEVIATION_PREPREPARE_SECS = old_value
+
+
+# TODO this test should actually fail someday when ts for PP is set
+# before node level processing (e.g. in zstack)
+def test_pp_obsolescence_check_fail_for_delayed(tdir, tconf,
+                                     looper,
+                                     txnPoolNodeSet,
+                                     sdk_pool_handle,
+                                     sdk_wallet_client):
+
+    delay = PATCHED_ACCEPTABLE_DEVIATION_PREPREPARE_SECS + 1
+    lagging_node = txnPoolNodeSet[-1]
+
+    # Prevent lagging node from ordering
+    with delay_rules(
+        lagging_node.nodeIbStasher, ppDelay(), pDelay(), cDelay()
+    ):
+        # Order request on all nodes except lagging one
+        sdk_send_random_and_check(looper, txnPoolNodeSet,
+                                  sdk_pool_handle, sdk_wallet_client, 1)
+        looper.run(asyncio.sleep(delay))
+
+    # Now delayed 3PC messages reach lagging node, so any transactions missed during
+    # catch up can be ordered, ensure that all nodes will have same data after that
+    ensure_all_nodes_have_same_data(looper, txnPoolNodeSet)
```

Comment:
We assume that all PPs will be rejected for the lagging node, don't we? If so, how the lagging node may have the same data?

Reply:
By requesting PP missed for Commits/Prepares